backend/src/route.rs

```diff
@@ -60,17 +84,23 @@ impl Router {
                 1.0
             };
             let cost = (penalty * road.cost_seconds() * 100.0) as usize;
-            match directions[&road.id] {
-                Direction::Forwards => {
-                    input_graph.add_edge(i1, i2, cost);
-                }
-                Direction::Backwards => {
-                    input_graph.add_edge(i2, i1, cost);
-                }
-                Direction::BothWays => {
-                    input_graph.add_edge(i1, i2, cost);
-                    input_graph.add_edge(i2, i1, cost);
-                }
+
+            let mut link_through_intersection =
+                |intersection: &Intersection, direction: Direction| {
+                    let from = node_map.get_or_insert((road.id, direction));
+                    for outgoing_road in intersection.allowed_movements_from(road.id, router_input)
+                    {
+                        let to = node_map.get_or_insert(outgoing_road);
```

Comment:
I almost got concerned here, but this is OK. In other projects using fast_paths, the initial contraction hierarchy (CH) gets rebuilt over time as the user edits something. Instead of building from scratch, it's sometimes faster to reuse the node ordering from the previous one. That requires the `NodeMap` to be exactly the same. For the representation in this PR, that would mean we need to carve out every (road, direction) pair in the NodeMap upfront.

But we're building from scratch every time (which is a perf issue we could improve on later), so for now, this is OK. Could be worth a brief comment here.

Reply:
Just to make sure I'm understanding - the comment should say something like:

> a given NodeId might refer to a different physical feature across rebuilds of the routing graph - do not assume they are stable

Is that what you mean?

Just to be clear, that behavior is not new to this PR.

[Previously](
https://github.com/a-b-street/ltn/pull/119/files/26665b3abfb3f3ecc60902ac63bdc0e79ade0028#diff-f2086e591ba396ff6f44c144fec9a8c43d1f5a44841c4cc098ee29b692fbd1c6L33), we skipped creating nodes when the road had a modal filter. If you added (or removed) a modal filter, that would similarly change the set of NodeIds.